substrate/client/network/sync/src/pending_responses.rs

```diff
@@ -79,6 +80,10 @@ impl<B: BlockT> PendingResponses<B> {
 			);
 			debug_assert!(false);
 		}
+
+		if let Some(waker) = self.waker.take() {
+			waker.wake();
```

Comment:
This requires that we called `poll` before. Not sure that this is removing any footgun.

Reply:
Looking at the code we are already polling `pending_responses` after we inserted any kind of element, so why is the task not waking up?

Reply:
> This requires that we called `poll` before. Not sure that this is removing any footgun.

Yeah, but we definitely called it at least once.

Reply:
> Looking at the code we are already polling `pending_responses` after we inserted any kind of element, so why is the task not waking up?

There is no problem with waking up right now, but it can shoot us once the polling in `SyncingEngine` is converted into `select!` and the order of inserting / polling `pending_responses` is not strictly defined.

Reply:
But for a `select!`. Once you insert data to the `pending_responses` your match arm will be finished and you start again from the top of the loop, aka calling `select!` again. This will poll all futures, including your `pending_responses` futures. So, not sure why we could miss there to poll it.